### PR TITLE
Feature/babel

### PR DESCRIPTION
**Pull request by emclaug2**
Changes proposed in this Pull Request:
- Remove all PropTypes from the React Library
- Use Babel to Transpile TS -> JS and generate PropTypes.
- Add build command to generate type definition files. (components/package.json)
- Swap awesome-typescript-loader with (ts-loader & ForkTsCheckerWebpackPlugin) for much faster storybook build times.

This is a typical build time for the storybook app on my desktop using awesome-typescript-loader.
Before refactor

This is the build time after the refactor.  ForkTsCheckerWebpackPlugin allows asynchronous type checking for faster builds.   If any type violations occur, `yarn build-storybook` still throws an exit code 1 in CircleCI. 
After

This is an example of the PropTypes being enforced in the Showcase project.  This PropType was generated using Babel. 
Example Prop Type Enforcement

### CODE REVIEW

**Comment by huayunh** [comment]
Should this be part of the PXB CLI too? If so we can log an issue there or somewhere in our backlog

**Comment by huayunh** [comment]
What was the reason to remove all the defaultProps and have them back in the object destructuring?

**Comment by joebochill** [comment]
This looks really clean and I like how we get a performance boost switching to babel. Did we check to make sure porptypes work for types that we extend? E.g., on the ScoreCard component, does the proptype checking work for props that come from the MUI Card (like `elevation`)?

**Comment by joebochill** [comment]
> Should this be part of the PXB CLI too? If so we can log an issue there or somewhere in our backlog

PX Blue CLI doesn't use PropTypes at all. Where were you thinking this would go?

**Comment by joebochill** [comment]
> What was the reason to remove all the defaultProps and have them back in the object destructuring?

From what I can see, using the auto-proptype generator plugin from babel doesn't support default props.

**Comment by emclaug2** [comment]
> > What was the reason to remove all the defaultProps and have them back in the object destructuring?
> 
> From what I can see, using the auto-proptype generator plugin from babel doesn't support default props.

It is entirely possible to re-add the defaultProps and remove them from the props destructuring;  I think it would be better to reduce the amount of duplicated type declaration.  

If we have a component that has all default props, we would have to list out these props in 3 places:
- Type declaration
- Props destructuring 
- defaultProps 

If we declare the defaultProps when we destructure, it doesn't look as pretty but it's less repetitive.

File: components/package.json [hunk 1]
```diff
@@ -4,7 +4,11 @@
   "description": "React components for PX Blue applications",
   "scripts": {
     "test": "jest src",
-    "build": "tsc --p tsconfig.lib.json",
+    "type-check": "tsc --noEmit",
+    "build:original": "tsc --p tsconfig.lib.json",
```
[the remark this answers is not in the document]
Comment (emclaug2) [reply components/package.json:8]: I am going to remove this command; I shouldn't have left this in.   It doesn't provide any value.

**Comment by emclaug2** [comment]
> This looks really clean and I like how we get a performance boost switching to babel. Did we check to make sure porptypes work for types that we extend? E.g., on the ScoreCard component, does the proptype checking work for props that come from the MUI Card (like `elevation`)?

Good question, I added a `elevation={'test'}` to one of the ScoreCards in the Showcase:
image

**Comment by joebochill** [comment]
> > This looks really clean and I like how we get a performance boost switching to babel. Did we check to make sure porptypes work for types that we extend? E.g., on the ScoreCard component, does the proptype checking work for props that come from the MUI Card (like `elevation`)?
> 
> Good question, I added `elevation={'test'}` to one of the ScoreCards in the Showcase:
> image
> 
> The performance boost is not from switching to babel; it takes about the same amount of time for us to build our `/components` folder. I switched the loaders we use when we build the storybook app because I was noticing some horrible performance issues.
> 
> Storybook was taking over two minutes to build on my mac and over a minute on my desktop, so I had to adjust our build strategy.

I see...babel is supposed to be faster than tsc when building components too, but since we use tsc to get the type declarations anyway, I guess we don't see the speed up there.

**Comment by joebochill** [comment]
> > > What was the reason to remove all the defaultProps and have them back in the object destructuring?
> > 
> > 
> > From what I can see, using the auto-proptype generator plugin from babel doesn't support default props.
> 
> It is entirely possible to re-add the defaultProps and remove them from the props destructuring; I removed defaultProps because I think it would be better to reduce the amount of duplicated type declaration.
> 
> If we have a component that has all default props, we would have to list out these props in 3 places:
> 
> * Type declaration
> * Props destructuring
> * defaultProps
> 
> If we set the defaultProps when we destructure, it doesn't look as pretty but it's less repetitive.

I agree with Evan...i prefer to limit ourselves to writing this stuff in as few places as possible (even though overall I like defaultProps).

**Comment by huayunh** [comment]
> > Should this be part of the PXB CLI too? If so we can log an issue there or somewhere in our backlog
> 
> PX Blue CLI doesn't use PropTypes at all. Where were you thinking this would go?

Both :) I'd love to see CLI going online ASAP.

**Pull request merged**